### PR TITLE
[Extension] oak-incremental-index: Low resource (RAM and CPU) incremental-index implementation using off-heap key/value map (OakMap)

### PR DESCRIPTION
**Pull request by liran-funaro**
Fixes [#9967](https://github.com/apache/druid/issues/9967).

### Description
This PR introduces an extension that improves Druid’s ingestion memory and CPU efficiency. It uses **60%** less memory and **50%** less CPU-time to achieve the same performance. This translated to nearly **double** the system's ingestion-throughput with the same memory budget, and a **75%** increase in throughput with the same CPU-time budget. The experimental setup and the full results are available [here](https://github.com/liran-funaro/druid/wiki/Evaluation).

To understand the motivation and rationale behind some of the proposed changes below, it is necessary to read the related issue: [#9967](https://github.com/apache/druid/issues/9967).

#### Introduce `OakIncrementalIndex`
We add a new incremental index implementation: `OakIncrementalIndex` as a Druid extension. The implementation is mostly borrowed from `OnheapIncrementalIndex` and `OffheapIncrementalIndex`, but has a few notable differences:
  1. It stores both **keys** and **values** off-heap (as opposed to the off-heap implementation that stores only the **values** off-heap)
  1. It is based on [`OakMap`](https://github.com/yahoo/Oak) instead of Java’s `ConcurrentSkipList` (CSL)
  1. It does not need to keep a mapping from row index to an actual row
  1. It is always ordered (as expected by `FactsHolder.persistIterable()`), even in plain mode
 
To achieve the best performance of our implementation, we had to refactor some interfaces of `IncrementalIndexRow` and `IncrementalIndex`. This refactoring is explained in #12122.

##### Key changed/added classes in this commit
 * Added everything under `druid/extensions-contrib/oak-incremental-index`. Most notable additions:
   - `OakIncrementalIndex`: follows the `IncrementalIndex` API
   - `OakIncrementalIndexRow`: follows the `IncrementalIndexRow` API
   - `OakKey`: handles the serialization, deserialization, and comparison of keys
 * Updated benchmarks to evaluate our implementation.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [X] added comments explaining the "why" and the intent of the code wherever it would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths.
- [X] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by liran-funaro** [comment]
Hi @jihoonson, @leventov, @gianm, @ebortnik, @b-slim, and @jon-wei.
Since you had an interest in our previous discussions regarding `OakIncrementalIndex`, we wanted to update you that our new and improved implementation (presented in this PR) reduces Druid's **CPU** and **RAM** consumption by over **50%** during the ingestion process. This translates to nearly **double** the system's ingestion-throughput with the same CPU and RAM budget. Please check out this PR and its related issue (#9967), as well as our detailed [system-level experiments](https://github.com/liran-funaro/druid/wiki/Evaluation).

**Comment by jihoonson** [comment]
Glad you guys are still working on this! I will take a look soon.

**Comment by liran-funaro** [comment]
@jihoonson Thank you. We are eager to hear your valuable feedback.

**Comment by liran-funaro** [comment]
@clintropolis We noticed your recent commits modified the same benchmarks as we did in this PR ([commit #a607c95](https://github.com/apache/druid/pull/10001/commits/a607c9565dafdca7614dcee2286a836b758d2a47)). Since you are familiar with this part of Druid, we will appreciate it if you can take the time to review the benchmark part of this PR. This commit alone can contribute to Druid since it resolves issues we had with the benchmarks. You can find a summary of the modifications [here](https://github.com/liran-funaro/druid/wiki/Code-Modifications#add-benchmark-cases-for-oakincrementalindex).

**Comment by liran-funaro** [comment]
We noticed a lot of Druid users run their workload on Amazon EC2. We want to point out that this PR will not only improve performance but will also reduce operational costs by allowing the users to choose more affordable EC2 instances without sacrificing performance.

The figure below shows the operational cost of different required ingestion throughput on [Amazon EC2](https://aws.amazon.com/ec2/pricing/on-demand/).
image

File: server/src/main/java/org/apache/druid/segment/realtime/plumber/Sink.java
Comment (yuanlihan) [inline server/src/main/java/org/apache/druid/segment/realtime/plumber/Sink.java:374]: The `maxBytesInMemory` is calculated from max heap memory by default while the oak incremental index will take it as available size of direct memory. It may lead to OOM error , see [here](https://github.com/apache/druid/issues/9967#issuecomment-645764620)

**Comment by liran-funaro** [comment]
Thanks to @yuanlihan for helping us find bugs on the realtime query side. Our system experiments focused on batch ingestion, so his contribution is highly appreciated.

File: server/src/main/java/org/apache/druid/segment/realtime/plumber/Sink.java
Comment (liran-funaro) [reply server/src/main/java/org/apache/druid/segment/realtime/plumber/Sink.java:374]: Hi @yuanlihan. I added a new commit which resolves this issue.
Now `maxBytesInMemory` default is set differently when using `oak` or `offheap` index type.
I also added this to the documentation.
Thanks again for identifying this issue!

**Comment by liran-funaro** [comment]
Hi @jihoonson, have you had a chance to check out our issue/PR? We will be happy to answer any questions you might have.

**Comment by liran-funaro** [comment]
### Updates
We are working with our production teams at Verizon Media toward testing our incremental-index implementation on actual production data. As part of this effort, we discovered some issues with: (1) sketches, and (2) scenarios where multiple indexes are used during ingestion in a single Peon.
We just updated the PR to solves these issues. Please let us know if you encounter similar issues (or others) and if this update solves these issues.

**Comment by liran-funaro** [comment]
We continue to evaluate Oak incremental-index on our (Verizon Media) production workload and seeing interesting results.
For example, we run Druid on two identical VMs that consumed events from the same Kafka feed for an entire day; one with Oak and one with the "vanilla" Druid (on-heap incremental index).

### Oak vs. on-heap ("vanilla") incremental-index comparison:
* Oak required 35% less flush operations for the entire duration
* Which produced 50% larger partitions
* As a result, Oak spent 50 **minutes** less time in flush operations; i.e., almost 40% reduction compared to the on-heap implementation
* This also reduced the time it took to merge and push the segments by almost a minute.

In the latest Druid Summit, we've seen that these kinds of optimizations are vital.
Companies invest resources in post-compaction, thus, creating larger partitions during ingestion can reduce these efforts cost. 

We appreciate any effort by the community for moving this forward, starting by reviewing our design proposal (#10321).

File: processing/src/main/java/org/apache/druid/segment/DoubleDimensionIndexer.java
Comment (Eshcar) [inline processing/src/main/java/org/apache/druid/segment/DoubleDimensionIndexer.java:135]: This change and the below changes are mainly for performance reasons, right?
to avoid the need to return all dims when they are off heap

**Review by Eshcar — APPROVED** [review]
LGTM
some minor comments inside

**Comment by wangxiaobaidu11** [comment]
i have merged your code in 0.18.1 branch，i am testing it

**Comment by liran-funaro** [comment]
> i have merged your code in 0.18.1 branch，i am testing it

I'm looking forward to hearing your results. If you need any help, please reach out.

**Comment by didip** [comment]
This looks like an amazing improvement. Can't wait for it to be merged in.

**Comment by liran-funaro** [comment]
Hi @a2l007, @jihoonson, @leventov, @gianm, @ebortnik, @b-slim, and @jon-wei.
We approciate your support and the time you invested so far on this extention.
After introducing a new extention-point in Druid for the incremental-index in previous PRs, now we are ready to proceed with including our extention as part of Druid.
We will approciate your review for this PR.

You are welcome to checkout the related issue (#9967), as well as our detailed [system-level experiments](https://github.com/liran-funaro/druid/wiki/Evaluation) for more information.

**Comment by clintropolis** [comment]
Apologies that it has been so slow getting traction on this PR, I have started to look things over and will try to do a full review soon.

**Comment by liran-funaro** [comment]
@clintropolis Did you have a chance to review this PR?
Is there any blockage? Anything I can help with?

**Comment by clintropolis** [comment]
>@clintropolis Did you have a chance to review this PR?
Is there any blockage? Anything I can help with?

Sorry I haven't yet, only blockage has been on my side for finding time to dig in, I think I should have some time later this week though.

**Comment by liran-funaro** [comment]
> > @clintropolis Did you have a chance to review this PR?
> > Is there any blockage? Anything I can help with?
> 
> Sorry I haven't yet, only blockage has been on my side for finding time to dig in, I think I should have some time later this week though.

Thanks!

**Review by a2l007 — COMMENTED** [review]
Thank you for your work on the PR! Please add user docs for the extension.
Left a couple of comments but haven't had a chance to do a detailed review. I'll defer to the other reviewers who are more experienced in this part of the code :)

File: extensions-contrib/oak-incremental-index/pom.xml
Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/pom.xml:32]: It updates automatically.

File: extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndex.java
Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndex.java:374]: I'm not sure why `CachingColumnSelectorFactory` is part of `OnheapIncrementalIndex` and not the parent class `IncrementalIndex`. I moved it to `IncrementalIndex`. I hope this will also make sense to the other reviewers.

File: extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexModule.java
Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexModule.java:45]: removed

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java:111]: Actually, this works. I added oak-incremental-index as a dependency to the benchmark module.
This is what I've seen other extensions have done.
Check out line 186 in `benchmarks/pom.xml`.

**Comment by liran-funaro** [comment]
Hi @clintropolis, @a2l007, and all.
It has been nearly a year since I opened this PR.
Although we had some progress merging #10335 and #10593, this PR remains without any significant review or any concrete plan towards merge.
I understand that this PR is big and that you are probably swamped with important work.
So to expedite the process, I would love to arrange an online meeting where I can guide you through the changes. Just pick the time.

PS
@a2l007, I'm planning to write the user doc once this PR will be on a trajectory toward merge. I don't want to invest more time and effort without any assurances that it wouldn't go to waste.

**Comment by liran-funaro** [comment]
Hi @a2l007, @jihoonson, @leventov, @gianm, @ebortnik, @b-slim, and @jon-wei.
I appreciate your support and the time you have committed to this extension.
To remove any blockers to the adoption of this improvement, I have published a new PR (#12122) that contains all the changes to Druid internals from this PR without the additional extension classes.
I hope that this will illustrate how this extension does not alter Druid's logic or behavior.
Therefore, it does not pose a threat to Druid stability and robustness and can be safely merged. 
Review PR #12122, and you will find that it is easy to read and review because there are no logical changes.

File: extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java
Comment (pjain1) [inline extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: this will throw exception with newer versions of Java, whats a good default in that case ?

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: It's a great question. In fact, I am not sure if the direct memory limit applies here at all.
Oak allocates its memory directly, so it is not subject to any JVM limitations (only to the OS memory limit).
So I'm not sure of a suitable default value.
An easy solution would be to throw an exception if this value is not set manually by the user (`maxBytesInMemory==0`).
Alternatively, rely on `maxRowsInMemory` to limit the usage and set the memory to be unlimited (same as `maxBytesInMemory==-1`).
@pjain1 @yurmix @a2l007 What do you think is the best approach here?

File: benchmarks/pom.xml
Comment (liran-funaro) [reply benchmarks/pom.xml]: Fixed

File: extensions-contrib/oak-incremental-index/pom.xml
Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/pom.xml]: Fixed

File: extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java
Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: @liran-funaro correct me if I am wrong when you say oak index is not subjected to any JVM limitations, isn't it subjected to Java's `XX:MaxDirectMemorySize` limit ?

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: No. It allocates the memory via `Unsafe.allocateMemory()`, which is not included in the direct memory size limitation.

Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: So in worst case oak off heap index can allocate more than the host memory limit and OOM killer will kill the process ?

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: Exactly

Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: In practice we set the Druid container memory limit to just a little more than Java max heap size + MaxDirectMemorySize (as per merge/query buffers config), I think its a common practice with those who run in containerized env to set both Max heap as well MaxDirectMemorySize limits to set container memory limit. 

So as per my understanding if Oak index in used in such scenario then I don't think there will be enough memory for it to work. I agree with your suggestion that we can have a middle manager config that can be set to indicate how much memory does oak index can use apart from Druid and other processes on the host. This number can be used as max bytes in memory to trigger a persist of in-memory segment to disk to prevent OOME. We should document this and say the memory limit should be more than `max heap size + MaxDirectMemorySize + oak index off heap limit + buffer for host processes`.

Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: A small correction, actually we do keep some amount of spare memory apart from what I have pointed out for caching segments on peons for faster querying, so Oak index can use that but will impact segment caching. Overall I think we should require setting memory limits for Oak index before enabling it as it has different memory characteristics and one will have to rethink the memory quotas allocated. @liran-funaro What do you think ?

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: I agree that the admin will have to re-configure its memory quotas when enabling this extension.

The on-heap index uses `Xmx/6` as a limitation (it uses `Xmx/3` overall because we can have two "live" indexes).
Since Oak doesn't need that heap space, the admin can set `Xms` to be `(2/3)*Xmx`, then the default Oak usage will be the same as the on-heap one.
I assume here that the JVM won't use more memory than needed (i.e., `Xms`).

So the extension's docs should suggest setting `Xms=(2/3)*Xmx`. @pjain1 @yurmix What do you think?

Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: Putting it in another way the docs can say that oak index is off heap so Xmx can be now be set to `2/3` of the earlier value. To keep memory requirements same please set max bytes in memory to `1/6` of Xmx ? 

How about `getDefaultMaxBytesInMemory` here return `JvmUtils.getRuntimeInfo().getMaxHeapSizeBytes() / 6` same as for `OnheapIncrementalIndex` so that default memory requirement is same as earlier when Xmx is set to `2/3` of earlier value ?

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: The problem with changing `Xmx` is that the ingestion task might be configured to use the on-heap index. Then it will have less memory for the index.
I think it is OK to rely on `Xms` because the Peon task has very little use of the heap.

Comment (pjain1) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: Ok this makes sense and keeps config backward compatible, but coming back to the point of this thread -
1. What should be the default value returned from `getDefaultMaxBytesInMemory` ? I think `JvmUtils.getRuntimeInfo().getMaxHeapSizeBytes() / 6` makes sense when Xms is set as per your recommendation.
2.  Docs should be make clear about oak memory characteristics and how to set host memory limits if `maxBytesInMemory` is set by user so that tasks do not fail with OOME.

Comment (liran-funaro) [reply extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndexSpec.java:93]: Agreed! I'll update the default value and the docs accordingly. Thanks!

**Comment by didip** [comment]
Please don't close this PR.

**Comment by exherb** [comment]
index performance is very critical for us.Too many tasks means too many segments.

**Comment by liran-funaro** [comment]
>Index performance is very critical for us. Too many tasks mean too many segments.

@exherb For batch ingestion, each task can ingest different time periods. Those can be merged when the task is done, similar to how we merge the results from periodic flushes. So we can benefit from better throughput without sacrificing the number of segments.

**Comment by exherb** [comment]
> > Index performance is very critical for us. Too many tasks mean too many segments.
> 
> @exherb For batch ingestion, each task can ingest different time periods. Those can be merged when the task is done, similar to how we merge the results from periodic flushes. So we can benefit from better throughput without sacrificing the number of segments.

It’s take about 3 hours to merge for us.

**Comment by liran-funaro** [comment]
> It takes about 3 hours to merge for us.

@exherb From my experiment, merging fewer, but larger segments take about the same amount of time as merging many small segments. Even so, this extension enables more data to be ingested into memory before it is flushed to disk, resulting in larger segments.

> Is this pull request ready?

@exherb Yes. Except for some conflicts that I can resolve (upon demand).

**Pull request closed without merging**
